h/streamer.py

```diff
@@ -590,76 +605,48 @@ def _authorized_to_read(effective_principals, annotation):
     return False
 
 
-def should_send_annotation_event(socket, annotation, event_data):
+def process_message(handler, reader, message):
     """
-    Inspects the passed annotation and action and decides whether or not
-    the underlying session should receive the event. If it should, the
-    action is wrapped up in a websocket packet and sent to the client.
-    """
-    if event_data['action'] == 'read':
-        return False
+    Deserialize and process a message from the reader.
 
-    if event_data['src_client_id'] == socket.client_id:
-        return False
+    For each message, `handler` is called with the deserialized message and a
+    single :py:class:`h.streamer.WebSocket` instance, and should return the
+    message to be sent to the client on that socket. The handler can return
+    `None`, to signify that no message should be sent, or a JSON-serializable
+    object. It is assumed that there is a 1:1 request-reply mapping between
+    incoming messages and messages to be sent out over the websockets.
 
-    if annotation.get('nipsa') and (
-            socket.request.authenticated_userid != annotation.get('user', '')):
-        return False
-
-    if not _authorized_to_read(
-            socket.request.effective_principals, annotation):
-        return False
-
-    # We don't send anything until we have received a filter from the client
-    if socket.filter is None:
-        return False
-
-    if not socket.filter.match(annotation, event_data['action']):
-        return False
-
-    return True
-
-
-def process_message(reader, message):
+    Any exceptions thrown by this function or by `handler` will be caught by
+    :py:class:`gnsq.Reader` and the message will be requeued as a result.
     """
-    Process a message from the queue.
+    data = json.loads(message.body)
 
-    Distributes messages from the queue to the appropriate handler functions.
-    """
-    if reader.topic == ANNOTATIONS_TOPIC:
-        broadcast_annotation_message(message, WebSocket.instances)
-    elif reader.topic == USER_TOPIC:
-        broadcast_session_change_message(message, WebSocket.instances)
-    else:
-        log.warn("streamer: don't know how to handle message from topic '%s'",
-                 reader.topic)
+    for socket in WebSocket.instances:
+        reply = handler(data, socket)
```

Comment:
`reply` seems an odd thing to call this (it's not being sent back to the thing that sent the original message)